### PR TITLE
Authorizations and validators

### PR DESCRIPTION
**Pull request by hrogel**

### CODE REVIEW

File: Extensions/StringExtension.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+//  Created by Héctor Rogel on 5/11/21.
+//
+
+import Foundation
+
+extension String {
+
+    func toDate() -> Date? {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "dd/MM/yyyy HH:mm"
```
Comment (ggdiez) [inline Extensions/StringExtension.swift:14]: The date formarter should be in a constant.

File: src/model/Authorization.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        nameReceive = ""
+        state = ""
+        sended = false
+        type = nil
+        initialDate = ""
+        endDate = "-"
+        observations = ""
+    }
+
+    init(name: String, type: AuthorizationType, initialDate: String, endDate: String, observations: String) {
```
Comment (ggdiez) [inline src/model/Authorization.swift:35]: You can use only one method with "" by default in all the items in the constructor except the value is passed.

File: src/model/Authorization.swift [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        id = ""
+        self.nameSend = name
+        nameReceive = ""
+        state = ""
+        sended = true
+        self.type = type
```
Comment (ggdiez) [inline src/model/Authorization.swift:41]: Why one assignations with self and other without?

File: src/ui/controllers/lists/configuration/ConfigurationDetailView/AuthorizationDetailViewController.swift [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        switch indexPath.row {
+        case 7:
+
+            let cell = tableView.dequeueReusableCell(withIdentifier: authorizationDetailButtonCellIdentifier, for: indexPath) as! AuthorizationDetailButtonCell
+            cell.configureCell(for: "Authorization_Detail_Accept".localized())
+            cell.button.addTarget(self, action: #selector(acceptButtonTapped(sender:)), for: .touchUpInside)
+            if authorization.state != "pending" || authorization.sended {
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationDetailView/AuthorizationDetailViewController.swift:116]: pending, accepted and revoked can be in a enumerated

File: src/ui/controllers/lists/configuration/ConfigurationDetailView/AuthorizationDetailViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    var errorMsgUpdated: ((String) -> Void)?
+    var errorMsg: String = "" {
+        didSet {
+            errorMsgUpdated?(errorMsg)
+        }
+    }
+
+    func acceptAuthorization(id: String) {
+        startConnection(id: id, code: 27)
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationDetailView/AuthorizationDetailViewModel.swift:33]: What is code 27? please use a constant.

File: src/ui/controllers/lists/configuration/ConfigurationDetailView/AuthorizationDetailViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    func rejectAuthorization(id: String) {
+        startConnection(id: id, code: 26)
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationDetailView/AuthorizationDetailViewModel.swift:37]: Same with 26.

File: src/ui/controllers/lists/configuration/ConfigurationDetailView/StateAuthorizationXMLController.swift [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
+        if elementName == "result" {
+            if isError {
+                auxResult = currentElementValue
+            } else {
+                dataSource = [currentElementValue.stringToBool() : ""]
+            }
+        }
+        if elementName == "errorMsg" {
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationDetailView/StateAuthorizationXMLController.swift:74]: Some "comparables" can be contants like "result" and "errorMsg"

File: src/ui/controllers/lists/configuration/ConfigurationView/AuthorizationCell.swift [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    @IBOutlet weak var date: UILabel!
+    @IBOutlet weak var state: UIImageView!
+    @IBOutlet weak var direction: UIImageView!
+
+    func configureCell(for authorization: Authorization) {
+        name.text = authorization.sended ? authorization.nameReceive : authorization.nameSend
+
+        date.text = authorization.endDate.toDate()?.toString() ?? "-"
+        switch authorization.state {
+        case "pending":
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationView/AuthorizationCell.swift:24]: Remember the enumerated for this states

File: src/ui/controllers/lists/configuration/ConfigurationView/AuthorizationXMLController.swift [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String : String] = [:]) {
+            if elementName == "auth" {
+                auxAuthorization = Authorization()
+                if let id = attributeDict["id"] {
+                    auxAuthorization.id = id
+                }
+                if let type = attributeDict["type"] {
+                    auxAuthorization.type = AuthorizationType(rawValue: type)
+                }
+                if let state = attributeDict["state"] {
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationView/AuthorizationXMLController.swift:44]: type, state, revdate... can be constants

File: src/ui/controllers/lists/configuration/ConfigurationView/AuthorizationXMLController.swift [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        newStr = newStr.replacingOccurrences(of: "&_lt", with: "<")
+        newStr = newStr.replacingOccurrences(of: "&_gt", with: ">")
+        if newStr == "\n" { return }
+
+        currentElementValue = newStr
+    }
+
+    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
+        switch elementName {
+        case "user":
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationView/AuthorizationXMLController.swift:72]: user ,authuser, observation and auth can be constants

File: src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewController.swift [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+        case true:
+            viewModel?.getAuthorizations()
+        case false:
+            viewModel?.getValidators()
+        }
+    }
+
+    // MARK: - Style and configurations
+    private func configureSegmentedControl() {
+        let color = UIColor.init(red: 105.0/255.0, green: 25.0/255.0, blue: 31.0/255.0, alpha: 1.0)
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewController.swift:51]: The color better in a colors file

File: src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    var areAuthorizationsPending: (() -> Void)?
+
+    func getAuthorizations() {
+        authorizationsFlag = true
+        authorizations = []
+        wsController?.delegate = self
+        let data = authorizationParser.buildRequest()
+        wsController?.loadPostRequest(withData: data, code: 24)
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewModel.swift:55]: What is code 24??

File: src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        wsController?.startConnection()
+    }
+
+    func getValidators() {
+        authorizationsFlag = false
+        validators = []
+        wsController?.delegate = self
+        let data = validatorParser.buildRequest()
+        wsController?.loadPostRequest(withData: data, code: 28)
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewModel.swift:64]: Same here with code 28. Better in a constant -> the name of the constant should be descriptive.

File: src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+        wsController?.startConnection()
+    }
+
+    func revokeValidator(id: String) {
+        isRevokingValidator = true
+        wsController?.delegate = self
+        let data = stateAuthorizationParser.buildRevokeValidatorRequest(id: id)
+        wsController?.loadPostRequest(withData: data, code: 30)
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewModel.swift:72]: Same here with 30

File: src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+        wsController?.startConnection()
+    }
+
+    private func pendingAuthorizations() {
+        for item in authorizations {
+            if item.state == "pending" {
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationView/ConfigurationViewModel.swift:78]: Could be pending be a enumerated?

File: src/ui/controllers/lists/configuration/ConfigurationView/ValidatorXMLController.swift [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        var mesg: String = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
+        mesg.append("<rqt/>")
+
+        return mesg
+    }
+}
+
+extension ValidatorXMLController: XMLParserDelegate {
+    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String : String] = [:]) {
+        if elementName == "user" {
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/ConfigurationView/ValidatorXMLController.swift:36]: user and id, better in a constants

File: src/ui/controllers/lists/configuration/CreateItem/CreateItemViewController.swift [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+        default:
+            return nil
+        }
+    }
+
+    private func getInitDate() -> String {
+        if (initDateField.text != "" && initTimeField.text != "") {
+        return "\(initDateField.text ?? "") \(initTimeField.text ?? "")"
+        } else {
+            return Date().utcDateToString(withFormat: "dd/MM/YYYY HH:mm")
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/CreateItem/CreateItemViewController.swift:156]: Use a constant for the format.

File: src/ui/controllers/lists/configuration/CreateItem/CreateItemViewModel.swift [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+    init(type: SearchType) {
+        self.type = type
+    }
+
+    func searchUser(string: String) {
+        isSearch = true
+        users = []
+        wsController?.delegate = self
+        let data = searchItemParser.buildRequest(string: string, searchType: type)
+        wsController?.loadPostRequest(withData: data, code: 19)
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/CreateItem/CreateItemViewModel.swift:52]: More code here, better in a descriptive constant.

File: src/ui/controllers/lists/configuration/CreateItem/SearchUserViewController.swift [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+        let cell = tableView.dequeueReusableCell(withIdentifier: userNameCellIdentifier, for: indexPath) as! UserNameCell
+        cell.configureCell(for: users[indexPath.row].name)
+        return cell
+    }
+}
+
+extension SearchUserViewController: UISearchBarDelegate {
+
+    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+        if searchText.count > 2 {
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/CreateItem/SearchUserViewController.swift:125]: 2 better in a constant.

File: src/ui/controllers/lists/configuration/CreateItem/SearchUserViewController.swift [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+            users = []
+            tableView.reloadData()
+        }
+    }
+
+    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
+        guard let text = searchBar.text else {
+            return
+        }
+        if text.count <= 3 {
```
Comment (ggdiez) [inline src/ui/controllers/lists/configuration/CreateItem/SearchUserViewController.swift:137]: Same here with 3

File: src/model/Authorization.swift (hunk 2 above)
Comment (hrogel) [reply src/model/Authorization.swift:35]: i don't understando, not all the items in this constructor have a default value

File: src/model/Authorization.swift (hunk 3 above)
Comment (hrogel) [reply src/model/Authorization.swift:41]: I use self when the parameter names match with the property names

Comment (ggdiez) [reply src/model/Authorization.swift:47]: ok, thanks for the answer

File: src/model/Authorization.swift (hunk 2 above)
Comment (ggdiez) [reply src/model/Authorization.swift:41]: ok, understand

**Pull request merged**